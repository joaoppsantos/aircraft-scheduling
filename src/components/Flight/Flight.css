.flight::-webkit-scrollbar {
    display: none;
}

.flight {
    height: calc(100vh - 15rem);
    width: calc(25vh);
    padding-top: 10px;
    background-color: lavender;
    margin-right: 20px;
    overflow-y: scroll;

    scrollbar-width: none; 
}

@media screen and (max-width: 670px) {
	.flight {
        width: 90%;
        margin: 0 auto;
	}
}
