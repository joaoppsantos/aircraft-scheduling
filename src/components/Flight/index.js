export { default as Flight } from './Flight';
