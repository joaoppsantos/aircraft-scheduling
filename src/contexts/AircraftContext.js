import { createContext, useEffect, useState } from 'react';
import request from '../request';

export const AircraftContext = createContext({});

export function AircraftProvider({ children }) {
	const [aircrafts, setAircrafts] = useState([]);
	const [isAircraftLoading, setIsAircraftLoading] = useState(true);
	const [currentAircraft, setCurrentAircraft] = useState(undefined);
	const [currentPercentage, setCurrentPercentage] = useState(0);

	const setPercentage = percentage => {
		setCurrentPercentage(percentage);
	};
	const selectAircraft = aircraft => {
		setCurrentAircraft(aircraft);
	};

	useEffect(() => {
		request.get(`/aircrafts?limit=5&offset=0`).then((response) => {
			setAircrafts(response.data.data);
			setIsAircraftLoading(false);
		});
	}, []);

	return (
		<AircraftContext.Provider value={{
			aircrafts,
			isAircraftLoading,
			currentAircraft,
			selectAircraft,
			currentPercentage,
			setPercentage
		}}>
			{children}
		</AircraftContext.Provider>
	)
}
