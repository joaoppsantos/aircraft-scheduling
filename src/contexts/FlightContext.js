import request from '../request';
import { createContext, useEffect, useState } from 'react';

export function FlightProvider({ children }) {
	const [flights, setFlights] = useState([]);
	const [activeFlights, setActiveFlights] = useState([]);
	const [offset, setOffset] = useState(-1);
	const [hasFlightsRemaining, setHasFlightsRemaining] = useState(true);
	const paginationLimit = 25;

	const toggleActiveFlight = (flight) => {
		let isFlightOnList = false;

		activeFlights.forEach((activeInfo) => {
			if (activeInfo.id === flight.id) {
				isFlightOnList = true;
			}
		});

		isFlightOnList ? 
			setActiveFlights(activeFlights.filter(activeInfo => activeInfo.id !== flight.id))
			: setActiveFlights([...activeFlights, flight]);
	}

	useEffect(() => {
		if (!hasFlightsRemaining) {
			return;
		}

		async function fetchData() {
			try {
				const newOffset = offset + 1;
				const response = await request.get(`/flights?limit=${paginationLimit}&offset=${newOffset}`);

				setOffset(newOffset);
				setFlights([...flights, ...response?.data?.data]);

				if (paginationLimit * (newOffset + 1) > response?.data?.pagination?.total) {
					setHasFlightsRemaining(false);
				}
			} catch (error) {
				console.log('Failed to load !');
			};
		}

		fetchData();
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [flights, hasFlightsRemaining]);

	return (
		<FlightContext.Provider value={{
			flights,
			toggleActiveFlight,
			activeFlights
		}}>
			{ children }
		</FlightContext.Provider>
	)
}

export const FlightContext = createContext({});