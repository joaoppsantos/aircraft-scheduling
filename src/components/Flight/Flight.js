import './Flight.css';
import { useContext } from 'react';
import { FlightContext } from '../../contexts/FlightContext';
import { FlightCard } from '../FlightCard';

const Flight = () => {
	const {
		flights
	} = useContext(FlightContext);

    return (
        <div className="flight">
            Flight
            {flights.map((flight, index) => {
                return <FlightCard key={index} {...flight} />
            })}
        </div>
    );
}

export default Flight;
