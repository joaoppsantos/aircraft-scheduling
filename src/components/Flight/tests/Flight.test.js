import * as React from 'react';
import { render } from '@testing-library/react';
import Flight from '../Flight';
import { FlightContext } from '../../../contexts/FlightContext';

const flightMock = [{
        "id": "AS1001",
        "departuretime": 21600,
        "arrivaltime": 26100,
        "readable_departure": "06:00",
        "readable_arrival": "07:15",
        "origin": "LFSB",
        "destination": "LFMN"
    },                 
    {
        "id": "AS1002",
        "departuretime": 27900,
        "arrivaltime": 32100,
        "readable_departure": "07:45",
        "readable_arrival": "08:55",
        "origin": "LFMN",
        "destination": "LFSB"
    }];

describe('<Flight />', () => {
    it('should render correctly', () => {
        const mockFlightContext = {
            flights: flightMock,
        };

        const { container } = render(
            <FlightContext.Provider value={mockFlightContext}>
                <Flight />
            </FlightContext.Provider>
        );

        expect(container).toMatchSnapshot();
    });
});
