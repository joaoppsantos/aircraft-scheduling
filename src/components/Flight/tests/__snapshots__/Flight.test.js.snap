// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flight /> should render correctly 1`] = `
<div>
  <div
    class="flight"
  >
    Flight
    <div>
      <div
        class="flight-card"
      >
        <span
          class="flight-id"
        >
          AS1001
        </span>
        <span>
          LFSB
           
          &gt;
           
          LFMN
        </span>
        <span>
          06:00
           
          &gt;
           
          07:15
        </span>
      </div>
    </div>
    <div>
      <div
        class="flight-card"
      >
        <span
          class="flight-id"
        >
          AS1002
        </span>
        <span>
          LFMN
           
          &gt;
           
          LFSB
        </span>
        <span>
          07:45
           
          &gt;
           
          08:55
        </span>
      </div>
    </div>
  </div>
</div>
`;
